pkg/containerruntimediscovery/types/cleanup.go

```diff
@@ -0,0 +1,63 @@
+// Copyright © 2023 Cisco Systems, Inc. and its affiliates.
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package types
+
+// Based on https://github.com/google/gvisor/blob/ba53672288fb/pkg/cleanup/cleanup.go
+
+// Cleanup allows defers to be aborted when cleanup needs to happen
+// conditionally. Usage:
+//
+//		 cu := cleanup.Make(func() { f.Close() })
+//		 defer cu.Clean() // failure before release is called will close the file.
+//		 ...
+//	   cu.Add(func() { f2.Close() })  // Adds another cleanup function
+//	   ...
+//		 cu.Release() // on success, aborts closing the file.
+//		 return f
+type Cleanup struct {
```

Comment:
note: `Cleanup` is not concurrently safe, not sure if it would be useful to call concurrently though (e.g. `Add` could definitely be called in parallel)

Reply:
Since is it no used concurrently I'd keep it as is. I'd expect the caller to ensure that either the `Cleanup` is wrapped with proper locking or `Cleanup` is extended with internal locking. We can add locking mechanism to the `Cleanup` struct later on, but I'd not extend the scope of the current PR with this.

Reply:
Lets just add that as a comment to the Cleanup struct for now